proof-systems/src/groth16/benches/bn382_gro16_test_circuits.rs

```diff
@@ -0,0 +1,248 @@
+use algebra::{
+    fields::bn_382::Fr,
+    curves::bn_382::Bn382,
+    UniformRand, PrimeField,
+};
+use r1cs_core::{SynthesisError, ConstraintSynthesizer, ConstraintSystem};
+use proof_systems::groth16::{generate_random_parameters, create_random_proof};
+
+use criterion::{BenchmarkId, BatchSize};
+use criterion::Criterion;
+use r1cs_std::Assignment;
+use r1cs_std::fields::fp::FpGadget;
+use r1cs_std::eq::EqGadget;
+use r1cs_std::fields::FieldGadget;
+use r1cs_std::alloc::AllocGadget;
+
+use rand::{
+    rngs::OsRng, thread_rng
+};
+
+use std::time::{SystemTime, UNIX_EPOCH};
+
+#[macro_use]
+extern crate criterion;
+
+#[macro_use]
+extern crate bench_utils;
+
+pub struct TestCircuit1<F: PrimeField> {
+    num_constraints: usize,
+    a: Option<F>,
+    b: Option<F>,
+}
+
+pub struct TestCircuit2<F: PrimeField> {
+    num_constraints: usize,
+    a: Option<F>,
+    b: Option<F>,
+}
+
+impl<F: PrimeField> ConstraintSynthesizer<F> for TestCircuit1<F> {
+    fn generate_constraints<CS: ConstraintSystem<F>>(
+        self,
+        cs: &mut CS,
+    ) -> Result<(), SynthesisError> {
+
+        let mut a_k_minus_1 = FpGadget::<F>::alloc_input(
+            cs.ns(|| "alloc a"),
+            || self.a.ok_or(SynthesisError::AssignmentMissing)
+        )?;
+
+        let mut b_k_minus_1 = FpGadget::<F>::alloc_input(
+            cs.ns(|| "alloc b"),
+            || self.b.ok_or(SynthesisError::AssignmentMissing)
+        )?;
+
+        let zero = FpGadget::<F>::zero(cs.ns(|| "alloc zero"))?;
+
+        a_k_minus_1.enforce_not_equal(cs.ns(|| "a_0 != 0"), &zero)?;
+        b_k_minus_1.enforce_not_equal(cs.ns(|| "b_0 != 0"), &zero)?;
+
+        for k in 0..(self.num_constraints - 5)/2 {
+
+            let a_k = FpGadget::<F>::alloc(
+                cs.ns(|| format!("alloc a_{}", k)),
+                || Ok(a_k_minus_1.value.get()? * &b_k_minus_1.value.get()?)
+            )?;
+
+            let b_k = FpGadget::<F>::alloc(
+                cs.ns(|| format!("alloc b_{}", k)),
+                || Ok(b_k_minus_1.value.get()? * &a_k_minus_1.value.get()?)
+            )?;
+
+            a_k_minus_1.mul_equals(
+                cs.ns(|| format!("a_{} * b_{} = a_{}", k - 1, k - 1, k)),
+                &b_k_minus_1,
+                &a_k
+            )?;
+
+            b_k_minus_1.mul_equals(
+                cs.ns(|| format!("b_{} * a_{} = b_{}", k - 1, k - 1, k)),
+                &a_k_minus_1,
+                &b_k
+            )?;
+
+            a_k_minus_1 = a_k;
+            b_k_minus_1 = b_k;
+        }
+
+        Ok(())
+    }
+}
+
+impl<F: PrimeField> ConstraintSynthesizer<F> for TestCircuit2<F> {
```

Comment:
Also here: let's comment the R1CS density d=1 and that every second constraint costs the synthesizer a field inversion (this should simulate EC arithmetics over prime fields).